### DISCUSSION TITLE
What is the largest sum you can make?

### DISCUSSION THREAD

**Discussion by ParanoidUser** [ParanoidUser/yolo#532] [Galaxy Trivia] [answered]
Starting in the bottom left corner and moving either up or right adding up the numbers along the way.

![image](https://user-images.githubusercontent.com/5120290/231196538-735cef68-adb3-4eb7-ba40-8da92f463c3b.png)

*answer in binary*

**Comment by peterbaccaro** [comment, marked as answer]
![puzzle](https://user-images.githubusercontent.com/74666079/231456935-8fe1b6ee-8347-47a6-b516-6d36cd852273.png)

1+3+5+3+2+4+5+4+1=28

In binary 28 is **11100**.

**Comment by Veercodeprog** [comment]
Answer is - 11010



// Define a 2D array
const array = [
    [2, 3, 5, 4, 1],
    [3, 2, 4, 3, 4],
    [5, 1, 3, 5, 2],
    [3, 3, 2, 3, 1],
    [1, 4, 2, 4, 4]
];

// Define variables to keep track of the current position and sum
let row = array.length - 1; // Starting from the bottom row
let col = 0; // Starting from the left column
let sum = array[row][col]; // Start with the value in the bottom left corner

// Loop until we reach the top row or rightmost column
while (row > 0 || col < array[0].length - 1) {
    // If we can move up, and the value above is greater than the value to the right
    if (row > 0 && (col === array[0].length - 1 || array[row - 1][col] > array[row][col + 1])) {
        row--; // Move up
    } else {
        col++; // Move right
    }
    sum += array[row][col]; // Add the current value to the sum
}

// Output the sum
console.log(sum.toString(2));;

**Reply by ParanoidUser** [reply]
Writing a program is a bright idea, but the algorithm does not look for the best possible solution.

**Reply by peterbaccaro** [reply]
Your code is deciding the direction to go based in the adjacent numbers only.
Instead the code needed to determine all valid routes and calculate the maximum sum possible.
Here's some Python code that will calculate the largest sum. 

```
from typing import List

def get_routes(grid: List[List[int]], row_idx: int, col_idx: int, route: List[int] = []) -> List[List[int]]:
    routes = []
    curr = grid[row_idx][col_idx]
    if row_idx == 0 and col_idx == len(grid[0]) - 1:
        routes.append(route)
    else:
        route.append(curr)
        if row_idx != 0:
            routes = routes + get_routes(grid, row_idx - 1, col_idx, route.copy())

        if col_idx != len(grid[0]) - 1:
            routes = routes + get_routes(grid, row_idx, col_idx + 1, route.copy())
    return routes


grid = [[2, 3, 5, 4, 1], [3, 2, 4, 3, 4], [5, 1, 3, 5, 2], [3, 3, 2, 3, 1], [1, 4, 2, 4, 4]]

# determine possible routes
routes = get_routes(grid, len(grid) - 1, 0)

# determine max sum possible from a route
max_sum_route = 0
for route in routes:
    sum_route = sum(route)
    max_sum_route = sum_route if sum_route > max_sum_route else max_sum_route

# print out the max sum in binary
print("{0:b}".format(max_sum_route))
```

**Reply by Veercodeprog** [reply]
function largestSum(grid) {
    const m = grid.length;
    const n = grid[0].length;
    const dp = new Array(m).fill(null).map(() => new Array(n).fill(0));
    dp[m - 1][0] = grid[m - 1][0];

    // initialize the last row with the sum of each element from the rightmost column to that row
    for (let j = 1; j < n; j++) {
        dp[m - 1][j] = grid[m - 1][j] + dp[m - 1][j - 1];
    }

    // initialize the first column with the sum of each element from the bottom row to that column
    for (let i = m - 2; i >= 0; i--) {
        dp[i][0] = grid[i][0] + dp[i + 1][0];
    }

    // fill the remaining cells in the dp table
    for (let i = m - 2; i >= 0; i--) {
        for (let j = 1; j < n; j++) {
            dp[i][j] = grid[i][j] + Math.max(dp[i + 1][j], dp[i][j - 1]);
        }
    }

    // return the largest sum, which is stored in the top-rightmost cell of the dp table
    return dp[0][n - 1];
}

const arr = [    [2, 3, 5, 4, 1],
    [3, 2, 4, 3, 4],
    [5, 1, 3, 5, 2],
    [3, 3, 2, 3, 1],
    [1, 4, 2, 4, 4]
];

console.log(largestSum(arr));

**Comment by Veercodeprog** [comment]
Modified my code to get the largest sum using tabulation to get the best possible time and space complexity.

**Comment by Veercodeprog** [comment]
is it necessary to consider target as grid[o][column-length], or we can move as we like to just reach the top row to get the largest sum.
for example like cases in
const arr = [    [204, -300, -500, 1, 1],
    [222, 2, 4, 3, 4],
    [45, 1, 3, 5, 2],
    [3, 3, 2, 3, 1],
    [1, 4, 2, 4, 4]
];

what will be the answer in this case ,largest sum would be when we reach grid[0][0]

**Reply by peterbaccaro** [reply]
The rules are that you can only move up or right.
For this new case the answer would be **100011101** which is 285 in decimal.

![puzzle-2](https://user-images.githubusercontent.com/74666079/231704368-b9bd41a5-c9c3-439b-bc84-41f5f3c2163d.png)

**Reply by Veercodeprog** [reply]
Is it necessary to reach top right position.like what if second and third columns is full of negative values.

**Reply by Veercodeprog** [reply]
also, check out https://github.com/Veercodeprog/Data-Structures-implementation/discussions for more questions and discussions

**Reply by peterbaccaro** [reply]
> Is it necessary to reach top right position.like what if second and third columns is full of negative values.

That's a valid point, the question didn't state that you needed to reach the top right corner although in the original question you needed to reach the top right corner as would be the case if all the numbers were positive.
In your example then, as you have added negative values the largest sum would therefore be 475 which would be 111011011 in binary.

**Comment by Veercodeprog** [comment]
we need more quiz questions like this
